### PR TITLE
Allow loading in parameters for the test device from fwupd.conf

### PR DESCRIPTION
**Pull request by hughsie**
for example:

    [test]
    RequestSupported=true
    RequestDelay=5000

Fixes https://github.com/fwupd/fwupd/issues/6836

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by superm1** [comment]
Can we strip some of the test plugin environment variables too with this kind of change?  

> @superm1 I was also pondering allowing something like fwupdmgr modify-config test:WriteSupported true -- sane?

Seems reasonable to be able to update variables in the namespace of a plugin.  Basically no namespace declared default to "fwupd" if one is declared, make sure it's in the list of stuff we allow.

**Comment by superm1** [comment]
> Can we strip some of the test plugin environment variables too with this kind of change?

Sorry I meant in the **documentation**.  `FWUPD_PLUGIN_TEST`

File: docs/fwupd.conf.md [hunk 1]
```diff
@@ -315,6 +315,34 @@ The `[thunderbolt]` section can contain the following parameters:
...
+
+  Delay in milliseconds to use when decompressing the test device.
+
+**WriteSupported={{test_Supported}}**
+
+  If the device write is supported.
+
+**WriteDelay={{test_DecompressDelay}}**
+
+  Delay in milliseconds to use when frobnicating the test device.
```
Comment (superm1) [inline docs/fwupd.conf.md:332]: don't use `frobnicating` here

File: docs/fwupd.conf.md [hunk 2]
```diff
@@ -315,6 +315,34 @@ The `[thunderbolt]` section can contain the following parameters:
...
+
+The `[test]` section can contain the following parameters:
+
+**DecompressDelay={{test_DecompressDelay}}**
+
+  Delay in milliseconds to use when decompressing the test device.
+
+**WriteSupported={{test_Supported}}**
+
+  If the device write is supported.
```
Comment (superm1) [inline docs/fwupd.conf.md:328]: maybe best to explain the expected behavior when this is marked unsupported

File: docs/fwupd.conf.md (hunk 1 above)
Comment (hughsie) [reply docs/fwupd.conf.md:332]: Hah, indeed. Thanks.

File: docs/fwupd.conf.md (hunk 2 above)
Comment (hughsie) [reply docs/fwupd.conf.md:328]: Done.

File: plugins/test/fu-test-plugin.c [hunk 3]
```diff
@@ -220,25 +157,38 @@ fu_test_plugin_write_firmware(FuPlugin *plugin,
...
-	FuTestPlugin *self = FU_TEST_PLUGIN(plugin);
 	const gchar *test = g_getenv("FWUPD_PLUGIN_TEST");
 	gboolean requires_activation = g_strcmp0(test, "requires-activation") == 0;
 	gboolean requires_reboot = g_strcmp0(test, "requires-reboot") == 0;
-	if (g_strcmp0(test, "fail") == 0) {
+	const gchar *tmp;
+	guint64 delay_decompress_ms = 0;
+	guint64 delay_write_ms = 0;
+	guint64 delay_verify_ms = 0;
+	guint64 delay_request_ms = 0;
```
Comment (superm1) [inline plugins/test/fu-test-plugin.c:167]: I guess this is asking for new test suite tests for the _timing_.  When > 0 check that nothing happened before the the time and check that something happened after the time.

File: plugins/test/fu-test-plugin.c [hunk 4]
```diff
@@ -362,7 +333,18 @@ static void
...
+
+static void
+fu_test_plugin_constructed(GObject *obj)
+{
+	FuPlugin *plugin = FU_PLUGIN(obj);
+	fu_plugin_set_config_default(plugin, "DecompressDelay", "0");
+	fu_plugin_set_config_default(plugin, "WriteSupported", "true");
+	fu_plugin_set_config_default(plugin, "WriteDelay", "0");
+	fu_plugin_set_config_default(plugin, "VerifyDelay", "0");
+	fu_plugin_set_config_default(plugin, "RequestSupported", "false");
```
[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/test/fu-test-plugin.c:346]: It's only wanted in one of the tests -- the others really don't want the interaction delay I think,

**Comment by hughsie** [comment]
>  FWUPD_PLUGIN_TEST

Yup, I think I can get rid of this; tomorrow tho, EOUTATIME for now.

File: docs/env.md [hunk 5]
```diff
@@ -41,7 +41,6 @@ Other variables, include:
 * `FWUPD_SOLOKEY_EMULATE` emulates a fake device for testing
 * `FWUPD_SUPERIO_DISABLE_MIRROR` disables the e-flash fixup to get byte-accurate hardware dumps
 * `FWUPD_SUPERIO_RECOVER` allow recovery of a corrupted SuperIO by hardcoding the device size
-* `FWUPD_TEST_PLUGIN_XML` used by the test plugin to load XML state out-of-band before startup
```
Comment (superm1) [inline docs/env.md:44]: get `FWUPD_PLUGIN_TEST` gone as well from line 38

File: src/fu-engine.c [hunk 6]
```diff
@@ -756,6 +757,12 @@ fu_engine_modify_config(FuEngine *self, const gchar *key, const gchar *value, GE
 			       "UpdateMotd",
 			       "UriSchemes",
 			       "VerboseDomains",
+			       "test:DecompressDelay",
+			       "test:WriteSupported",
+			       "test:WriteDelay",
+			       "test:VerifyDelay",
+			       "test:RequestSupported",
+			       "test:RequestDelay",
```
Comment (superm1) [inline src/fu-engine.c:765]: add to bash-completion too please

Comment (hughsie) [reply src/fu-engine.c:765]: Done.

File: docs/env.md (hunk 5 above)
Comment (hughsie) [reply docs/env.md:44]: variable 100% gone!

File: plugins/test/fu-test-plugin.c (hunk 3 above)
Comment (hughsie) [reply plugins/test/fu-test-plugin.c:167]: I'm hesitant to do this in the tests; we have to cope with CI doing things with huge amount of skew, and also it slows the test down considerably even for 10ms delays.

File: docs/fwupd.conf.md [hunk 7]
```diff
@@ -315,6 +315,54 @@ The `[thunderbolt]` section can contain the following parameters:
...
+
+  If the device needs activating before deploying the update.
+
+**CompositeChild={{test_CompositeChild}}**
+
+  If the device should have a child device.
+
+**AnotherWriteRequired={{test_AnotherWriteRequired}}**
+
+  Do two passes of the write function.
```
Comment (superm1) [inline docs/fwupd.conf.md:364]: I think these should be sorted

**Review by superm1 — COMMENTED** [review]
oooh I like the reset defaults.  Can we bubble that up to the tools too?  It would be nice to have a reset knob for tools too I think.

**Comment by hughsie** [comment]
>  Can we bubble that up to the tools too

Something like `fwupdtool reset-config section` -- where section is `fwupd`, `redfish` etc?

**Comment by hughsie** [comment]
> we can add the sections as bash completion targets as needed too

I always suck at that; could you do the bash completion bits and I can do the fwupd**tool** bits?

**Comment by superm1** [comment]
> > we can add the sections as bash completion targets as needed too
> 
> I always suck at that; could you do the bash completion bits and I can do the fwupd**tool** bits?

yeah go ahead and push it how you want to do it and I'll follow it up.

File: docs/fwupd.conf.md (hunk 7 above)
Comment (hughsie) [reply docs/fwupd.conf.md:364]: Agree, fixed.

**Review by superm1 — COMMENTED** [review]
Inconsistent between 1.9.14 and 1.9.15 in good.

**Comment by hughsie** [comment]
> Inconsistent between 1.9.14 and 1.9.15 in good.

I was thinking of releasing 1.9.14 with the contents of 1_9_X  -- and then post tag backport this.

**Comment by superm1** [comment]
Makes sense, just update them all to 1.9.15 then in gidoc

**Comment by hughsie** [comment]
Ohh i see what you mean; fixed.

**Review by superm1 — APPROVED** [review]

**Pull request merged**